megamek/src/megamek/client/bot/princess/PathRanker.java

```diff
@@ -105,18 +107,40 @@ abstract RankedPath rankPath(MovePath path, IGame game, int maxRange, double fal
             final BigDecimal numberPaths = new BigDecimal(validPaths.size());
             BigDecimal count = BigDecimal.ZERO;
             BigDecimal interval = new BigDecimal(5);
+            
+            boolean pathsHaveExpectedDamage = false;
+            
             for (MovePath path : validPaths) {
                 count = count.add(BigDecimal.ONE);
                 
-                returnPaths.add(rankPath(path, game, maxRange, fallTolerance, startingHomeDistance, enemies,
-                                         allyCenter));
+                RankedPath rankedPath = rankPath(path, game, maxRange, fallTolerance, startingHomeDistance, enemies,
+                        allyCenter);
+                returnPaths.add(rankedPath);
+                
+                // we want to keep track of if any of the paths we've considered have some kind of damage potential
+                pathsHaveExpectedDamage |= (rankedPath.getExpectedDamage() > 0);
+                
                 BigDecimal percent = count.divide(numberPaths, 2, RoundingMode.DOWN).multiply(new BigDecimal(100))
                                           .round(new MathContext(0, RoundingMode.DOWN));
                 if (percent.compareTo(interval) >= 0) {
                     getOwner().sendChat("... " + percent.intValue() + "% complete.", LogLevel.INFO);
                     interval = percent.add(new BigDecimal(5));
                 }
             }
+            
+            Entity mover = movePaths.get(0).getEntity();
+            UnitBehavior behaviorTracker = getOwner().getUnitBehaviorTracker();
+            boolean noDamageButCanDoDamage = !pathsHaveExpectedDamage && (FireControl.getMaxDamageAtRange(mover, 1, false, false) > 0);
+            
+            // if we're trying to fight, but aren't going to be doing any damage no matter how we move
+            // then let's try to get closer
+            if(noDamageButCanDoDamage && behaviorTracker.getBehaviorType(mover, getOwner()) == BehaviorType.Engaged) {
+                
+                behaviorTracker.overrideBehaviorType(mover, BehaviorType.MoveToContact);
+                return rankPaths(getOwner().getMovePathsAndSetNecessaryTargets(mover, true), game, maxRange, fallTolerance, 
```

Comment:
Should their behavior change back?

Reply:
No, that would result in an endless loop. The idea here is that if none of the considered paths have any damage potential whatsoever, we should move towards the nearest enemy unit instead.